components/raftstore/src/store/peer.rs

```diff
@@ -3823,6 +3903,83 @@ where
         Ok(prs)
     }
 
+    pub fn maybe_witness_transfer_leader<T>(&mut self, ctx: &PollContext<EK, ER, T>) {
+        if !(self.is_witness() && self.is_leader()) {
+            return;
+        }
+        fail_point!("witness_transfer_leader", |_| {});
+        info!(
+            "witness try to transfer leader";
+            "region_id" => self.region_id,
+            "peer_id" => self.peer.get_id(),
+        );
+
+        if self.raft_group.raft.lead_transferee.is_some() {
+            info!("witness already transferring leader");
+            // already being in transfer leader
+            return;
+        }
+
+        let prs = self.raft_group.raft.prs();
+        // find other peers that are not down
+        let (_, peer) = self
+            .region()
+            .get_peers()
+            .iter()
+            .filter(|peer| peer.id != self.peer.id)
+            .filter(|peer| !peer.is_witness)
```

Comment:
after https://github.com/tikv/tikv/pull/13209 has been merged, should we add a filter to ignore non_witnesses who are waiting for data?

Reply:
yes, and seems should filter out learner as well

Reply:
add a todo here